server-spi/src/main/java/org/keycloak/models/UserModel.java

```diff
@@ -143,6 +158,11 @@
 
     String getEmail();
 
+    /**
+     * Sets email for this user, email should be switched to lowercase before setting.
```

Comment:
This is because JPA layer is doing it [this way](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java#L331). 

If we decide to change this to not use lowercase, we should probably do this in separate Jira across whole codebase.

Reply:
JPA is doing it this way to ensure that letter case is ignored for usernames. However, it should not be mandated to be done on the level of the storage. It should only be mandated that search ignores the letter case.